ocaml/xapi/repository_helpers.ml

```diff
@@ -0,0 +1,778 @@
+      update_id = (member "updateId" j |> to_string_option)
+    }
+end
+
+module Guidance = struct
+  type t =
+    | RebootHost
+    | RestartToolstack
+    | EvacuateHost
+    | RestartDeviceModel
+
+  type guidance_kind = Absolute | Recommended
+
+  let to_string = function
+    | RebootHost -> "RebootHost"
+    | RestartToolstack -> "RestartToolstack"
+    | EvacuateHost -> "EvacuateHost"
+    | RestartDeviceModel -> "RestartDeviceModel"
+
+  exception Unknown_guidance
+
+  let of_string = function
+    | "RebootHost" -> RebootHost
+    | "RestartToolstack" -> RestartToolstack
+    | "EvacuateHost" -> EvacuateHost
+    | "RestartDeviceModel" -> RestartDeviceModel
+    | _ -> raise Unknown_guidance
+
+  let of_xml = function
+    | Xml.Element ("guidance", _, [Xml.PCData v]) ->
+      of_string v
+    | _ ->
+      error "Unknown node in <absolute|recommended_guidances>";
+      raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+
+  let set_of_list l =
+    l
+    |> List.map to_string
+    |> GuidanceStrSet.of_list
+
+  let eq l s = GuidanceStrSet.equal (set_of_list l) (set_of_list s)
+
+  let eq_set1 = eq [EvacuateHost; RestartToolstack]
+
+  let eq_set2 = eq [RestartDeviceModel; RestartToolstack]
+
+  let eq_set3 = eq [RestartDeviceModel; EvacuateHost]
+
+  let eq_set4 = eq [EvacuateHost; RestartToolstack; RestartDeviceModel]
+
+  let assert_valid_guidances = function
+    | [RebootHost]
+    | [EvacuateHost]
+    | [RestartToolstack]
+    | [RestartDeviceModel] -> ()
+    | l when eq_set1 l ->
+      (* EvacuateHost and RestartToolstack *)
+      ()
+    | l when eq_set2 l ->
+      (* RestartDeviceModel and RestartToolstack *)
+      ()
+    | l when eq_set3 l ->
+      (* RestartDeviceModel and EvacuateHost *)
+      ()
+    | l when eq_set4 l ->
+      (* EvacuateHost, RestartToolstack and RestartDeviceModel *)
+      ()
+    | l ->
+      let msg =
+        Printf.sprintf "Found wrong guidance(s) before applying updates: %s"
+          (String.concat ";" (List.map to_string l))
+      in
+      raise Api_errors.(Server_error (internal_error, [msg]))
+
+  let resort_guidances gs=
+    match GuidanceStrSet.find_opt "RebootHost" gs with
+    | Some _ -> GuidanceStrSet.singleton "RebootHost"
+    | None -> gs
+end
+
+module Applicability = struct
+  type inequality = Lt | Eq | Gt | Lte | Gte
+
+  type order = LT | EQ | GT
+
+  let string_of_order = function
+    | LT -> "<"
+    | EQ -> "="
+    | GT -> ">"
+
+  type t = {
+      name: string
+    ; arch: string
+    ; inequality: inequality option
+    ; epoch: string
+    ; version: string
+    ; release: string
+  }
+
+  type segment_of_version =
+    | Int of int
+    | Str of string
+
+  exception Invalid_inequality
+
+  let string_of_inequality = function
+    | Some Lt -> "<"
+    | Some Eq -> "="
+    | Some Gt -> ">"
+    | Some Gte -> ">="
+    | Some Lte -> "<="
+    | _ -> "InvalidInequality"
+
+  let inequality_of_string = function
```

Comment:
Why do we need to ever convert from strings like "gte" to an internal representation? We are parsing strings like "<=" and I understand that this need error handling. But I don't understand how strings like "gte" can arise.

Reply:
`inequality_of_string` is used in `of_xml`, which implies these strings appear in xml files we are trying to parse ... is that right @minglumlu ?

I think a unit test would really help here. We can show an example of the sort of xml file we want to be able to parse, and prove that we are parsing it correctly

Reply:
If that's the case, then some error handling is justified. However, I was expecting that XML decoding would not expose us to strings like "gte" if they are coming from ">=" originally. Overall, this still looks convoluted to me but i'm happy to move on and close this PR.

Reply:
The value in XML is designed as "gte", "lte", "lt" or "gt".  The string like ">=" is used only for logging.
And a following PR will have all unit tests against all of these.

Reply:
I changed the string ">=" to "gte" for consistency.